### DISCUSSION TITLE
Tiled custom property collisionType: how to get object data on collisionStart?

### DISCUSSION THREAD

**Discussion by cdelstad** [excaliburjs/Excalibur#2847] [Q&A] [answered]
Hey folks, I am just beginning my Excalibur journey and am migrating my existing Phaser.js project to Excalibur. So I may be missing something quintessential.
The Tile Plugin can add a collision in an Object Layer (custom property collisionType=passive). However, when the collision happens on the player (collisionStart event), I cannot find any information on the TileObject. I see an 'other' object as part of the collision, and it says it is an actor, however, it does not give me an object reference or object data so I can tell what object it is. I cannot seem to add a name to the object when it is added by the Tiled Plugin. I also cannot drill down to find object information such as a custom property like name = 'portal.' Adding an object.on('collisionStart'... doesn't work because it is a TileObject and doesn't have a body.
Is there a way to get tile object properties when the actor collides with it? My point of reference is Phaser which gives you a pointer to the object so you can access the object directly to get properties or directly manipulate it, such as killing the object or subtracting hit points, etc.
My preference would be to be able to loop through each object in the object layer from Tiled and add a collisionStart callback individually to do specific things - or other collision events such as collisionEnd, et al.
I could loop through the objects on scene load, and create an Actor at the same X/Y location with a circle collider, and then add a callback on that actors collisionStart event, however then it defeats the purpose of having the Tiled Plugin add a collisionType. I would also have double objects and may need to remember to destroy both.
In my example, a Trigger may be a more appropriate pattern, but I do not see a way to add a trigger with an action to teleport a player to a different map via the Tiled Plugin.
Has anyone experienced something similar, and how did you solve it? 
Thanks in advance,
-Chad

**Comment by eonarheim** [comment, marked as answer]
Hi @cdelstad 

Apologies it took a moment to answer you!

> The Tile Plugin can add a collision in an Object Layer (custom property collisionType=passive). However, when the collision happens on the player (collisionStart event), I cannot find any information on the TileObject. I see an 'other' object as part of the collision, and it says it is an actor, however, it does not give me an object reference or object data so I can tell what object it is. I cannot seem to add a name to the object when it is added by the Tiled Plugin. 

In the current incarnation of the plugin, this can be accessed off of a component on the colliding actors. (This doesn't seem to be very well documented unfortunately something I've added to my list of docs to fix)

```typescript

playerActor.on('collisionstart', evt => {
    const data = evt.other.get(TiledObjectComponent)
    console.log(data);
});
```

Giving Tiled objects a Name property also propagates through to the Actor as well.

<img width="500" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/612071/34b25726-cd43-4e55-855e-62d769b0f12c">


Additionally the Tiled editor renamed Type to Class so this get's confusing! These properties are needed (unfortunately casing also matters in the current incarnation for CollisionType = Passive to work)

![image](https://github.com/excaliburjs/Excalibur/assets/612071/a9c34e91-0f70-4934-82e8-fb748c3e1b79)


> I could loop through the objects on scene load, and create an Actor at the same X/Y location with a circle collider, and then add a callback on that actors collisionStart event, however then it defeats the purpose of having the Tiled Plugin add a collisionType. 

This also works, and it looks like you might have to do that for Circle colliders because the current automagic integration appears to have a bug and the sizing/position is wrong :( 

<img width="256" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/612071/c667a63e-0f52-449d-a054-0e76fd395b18">

<img width="351" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/612071/65c2e631-5bc6-41d5-9f31-87246deebfec">

Here is the workaround code I shared with you in Discord

```typescript
    const script = Resources.TiledMap.data.getObjectLayerByName("script");

    for (let obj of script.objects) {
        // Cannot rely on [0] so need to loop for value = portal and name = name?
        if (obj.properties.some(p => p.value ==='portal')){ // < -- no longer depends on prop order
            // Actors have a built in circle collider if radius is set
            const actorWithCircleCollider = new ex.Actor({
                pos: ex.vec(obj.x + obj.width!/2, obj.y + obj.width!/2), // < -- tiled renders stuff oddly
                radius: (obj.width || 20) / 2, // < - radius is 1/2 width
                collisionType: ex.CollisionType.Passive,
                color: ex.Color.Green, // <-- debug color
                z: 99,  // < -- debug z above all
                name:'portal',
                anchor: ex.vec(0.5, 0.5), // < -- switch back to center
            });

            actorWithCircleCollider.on('collisionstart', () => {
                console.log('Look at the portal on that!');
            });
            game.currentScene.add(actorWithCircleCollider);
        }
    }
```


I've 100% fixed this in the new plugin PR (although unmerged) https://github.com/excaliburjs/excalibur-tiled/pull/477

**Reply by eonarheim** [reply]
Switching from circles to boxcollider does seem to work OOTB as a workaround

<img width="544" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/612071/087809dc-89ec-4d76-a539-863683766cff">
<img width="176" alt="image" src="https://github.com/excaliburjs/Excalibur/assets/612071/598d9bb9-5113-4767-910d-6579b1974807">
